### PR TITLE
Add Functionality AllReviewsScreen

### PR DESCRIPTION
**Pull request by LouisJamesVasseur**
### What is this PR?  

This PR aims to give users greater agency with what reviews they can add/see on the app, by adding to the ways that they can interact with their own reviews and others'

### Why? 

It's important, for the M2 release of our app to start giving our users a complete experience, and reviews are a cornerstone of any community-based app. With this PR, we now have a complete and functional version of the Reviewing system

### How: 
A few things were implemented:
- A delete button was added, so users can now delete the review they've added on a Parking
- Two basic sorting methods (by Rating and by Date) so Users can see the reviews most pertinent to their criteria faster
- An overhaul of the UI of the Screen to match the new features

Screenshot 2024-11-13 at 00 29 29
Screenshot 2024-11-13 at 00 29 39

### Tests

AllReviewsScreen tests were adapted to the new screen functions. The new updateReviewScore functions were given new tests as well

### Code Coverage

Screenshot 2024-11-13 at 17 08 22

### CODE REVIEW

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt [hunk 1]
```diff
@@ -290,6 +264,44 @@ class ParkingViewModel(
...
+        },
+        onFailure = {})
+  }
+
+  fun handleReviewUpdate(
+      parking: Parking = selectedParking.value!!,
+      newScore: Double,
+      oldScore: Double
+  ) {
+    val delta = if (parking.nbReviews != 0) (newScore - oldScore) / parking.nbReviews else 0.0
```
Comment (sean-prz) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:300]: We shouldn't `if else` on the nbReviews. 
If it is zero and the user can modify his review, something has gone. 
Making the app crash with an exception might be overkill, but using `Log.e` seems reasonable and then instantly returning without running the code

**Review by sean-prz — COMMENTED** [review]
I am impressed, the UI is so much cleaner and everything seems to work on my phone.
There is just the edge case when there a spot has more reviews than my screen can display that I can't test, so I don't know if the lazy col. works correctly. 

Just address my comments about the code and re-request my review when it's done.

**Comment by sean-prz** [comment]
image
I ran the jacocoTestReport on your branch

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 2]
```diff
@@ -53,176 +64,297 @@ fun Timestamp?.toFormattedDate(): String {
...
+                .clickable { onOptionSelected(ReviewSortingOption.Rating) }
+                .background(
+                    color =
+                        if (selectedSortingOption == ReviewSortingOption.Rating)
+                            MaterialTheme.colorScheme.secondaryContainer
+                        else Color.Transparent,
+                    shape = MaterialTheme.shapes.small)
+                .padding(12.dp)) {
+          Text(
+              text = "Sort By Rating",
```
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:155]: This is a really good idea, I should've thought of that!

File: app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt [hunk 3]
```diff
@@ -43,6 +43,13 @@ class ReviewViewModel(private val reviewRepository: ReviewRepository) : ViewMode
         { Log.e("ReviewViewModel", "Error getting reviews", it) })
   }
 
+  fun deleteReviewById(review: Review) {
+    reviewRepository.deleteReviewById(
+        review.uid,
+        { Log.e("ReviewViewModel", "REVIEW DELETED") },
```
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:49]: Oops, shouldn't have been left in, this was for debugging purposes

[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/review/ReviewViewModel.kt:49]: (if you look, the message it sends isn't even SUPPOSED to be an error, Review Deleted is the intended onSuccess behavior!)

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt [hunk 4]
```diff
@@ -290,6 +264,44 @@ class ParkingViewModel(
...
+    parking.nbReviews -= 1
+    parkingRepository.updateParking(parking, {}, {})
+  }
+
+  fun handleNewReview(parking: Parking = selectedParking.value!!, newScore: Double) {
+    parking.avgScore =
+        (100 * ((parking.avgScore * parking.nbReviews) + newScore) / (parking.nbReviews + 1))
+            .toInt() / 100.00
+    parking.nbReviews += 1
+    parkingRepository.updateParking(
```
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:283]: Good point

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt (hunk 1 above)
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:300]: I agree here, but not in the deletion logic. In that one, the number of parkings being 0 makes sense. I will update here though

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 5]
```diff
@@ -53,176 +64,297 @@ fun Timestamp?.toFormattedDate(): String {
...
+                .padding(vertical = 4.dp)
+                .clickable { onOptionSelected(ReviewSortingOption.DateTime) }
+                .background(
+                    color =
+                        if (selectedSortingOption == ReviewSortingOption.DateTime)
+                            MaterialTheme.colorScheme.secondaryContainer
+                        else Color.Transparent,
+                    shape = MaterialTheme.shapes.small)
+                .padding(12.dp)) {
+          Text(
```
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:175]: Indeed, thanks!

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt [hunk 6]
```diff
@@ -290,6 +264,44 @@ class ParkingViewModel(
     }
   }
 
+  fun handleReviewDeletion(parking: Parking = selectedParking.value!!, newScore: Double) {
```
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt]: Addressed in the most recent commit (also agree that the naming might have not been the best)

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -114,6 +114,8 @@
     <string name="all_review_rating">Rating: %.1f</string>
     <string name="all_review_text">Text: %s</string>
     <string name="undefined_username">Unknown user</string>
+    <string name="sort_by_rating">Sort By Rating (Best Rated First)</string>
+    <string name="sort_by_date">Sort By Date (Most Recent First)</string>
```
Comment (Armetiis) [inline app/src/main/res/values/strings.xml:118]: this is most likely just nitpicking but I am uncomfortable with capping the begining of each word when it's not necessary

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 8]
```diff
@@ -53,176 +64,297 @@ fun Timestamp?.toFormattedDate(): String {
...
+    onToggle: () -> Unit,
+    content: @Composable () -> Unit
+) {
+  Column(
+      modifier =
+          Modifier.padding(8.dp)
+              .border(1.dp, Color.Gray, shape = MaterialTheme.shapes.medium)
+              .background(MaterialTheme.colorScheme.surface, shape = MaterialTheme.shapes.medium)
+              .padding(8.dp)) {
+        com.github.se.cyrcle.ui.theme.atoms.Text(
```
[the remark this answers is not in the document]
Comment (Armetiis) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:85]: Now honestly I don't think this title is needed in the first place.

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 9]
```diff
@@ -53,176 +64,297 @@ fun Timestamp?.toFormattedDate(): String {
...
+                        else Color.Transparent,
+                    shape = MaterialTheme.shapes.small)
+                .padding(12.dp)) {
+          Text(
+              text = stringResource(R.string.sort_by_rating),
+              style = MaterialTheme.typography.bodyMedium,
+              color =
+                  if (selectedSortingOption == ReviewSortingOption.Rating)
+                      MaterialTheme.colorScheme.primary
+                  else MaterialTheme.colorScheme.onSurface)
```
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:160]: here don't you think we can keep the text White in both cases and use cerulean rather than secondary container? 

Not an urgent matter, in particular with the future theme refactor

**Review by Armetiis — COMMENTED** [review]
Really good job on improving both the backend and the screen. They are much more solid and beautiful, respectfully. I just left you some comments on the code. 
Most of the reviews I left in the screen are nitpicks. The only thing that's bothering me would be the documentation in the ParkingViewModel.  Either way they are quick fixes and shouldn't cause too much trouble.

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt (hunk 9 above)
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:160]: I agree that this can and should be handled in the theme refactor, as changes I make on this now would just be overwritten

File: app/src/main/res/values/strings.xml (hunk 7 above)
Comment (LouisJamesVasseur) [reply app/src/main/res/values/strings.xml:118]: This is a stylistic choice, and one that respects conventions to my knowledge, so this should probably be handled at a later time

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt [hunk 10]
```diff
@@ -53,176 +64,297 @@ fun Timestamp?.toFormattedDate(): String {
...
-                                              style = MaterialTheme.typography.bodySmall,
-                                              maxLines = 2,
-                                              modifier = Modifier.testTag("ReviewText$index"))
-                                        }
+
+                                    Text(
+                                        text = "By: ${uidOfOwner.value}",
+                                        fontWeight = FontWeight.Medium,
+                                        color = MaterialTheme.colorScheme.onSecondaryContainer,
+                                        style = MaterialTheme.typography.bodySmall,
```
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:276]: This can also be addressed after the style refactor

File: app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/ui/review/AllReviewsScreen.kt:85]: Will do that, it indeed will make the code cleaner

**Review by Armetiis — APPROVED** [review]
Great job. One step closer to a fully functional casino.. ehem I meant parking app
